extensions/amp-next-page/0.1/next-page-service.js

```diff
@@ -0,0 +1,374 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {CSS} from '../../../build/amp-next-page-0.1.css';
+import {MultidocManager} from '../../../src/runtime';
+import {PositionObserverFidelity} from '../../../src/service/position-observer/position-observer-worker';
+import {Services} from '../../../src/services';
+import {dev} from '../../../src/log';
+import {getAmpdoc, getServiceForDoc} from '../../../src/service';
+import {
+  installPositionObserverServiceForDoc,
+} from '../../../src/service/position-observer/position-observer-impl';
+import {installStylesForDoc} from '../../../src/style-installer';
+import {layoutRectLtwh} from '../../../src/layout-rect';
+import {parseUrl} from '../../../src/url';
+import {setStyle} from '../../../src/style';
+import {triggerAnalyticsEvent} from '../../../src/analytics';
+
+// TODO(emarchiori): Make this a configurable parameter.
+const SEPARATOR_RECOS = 3;
+
+const MAX_ARTICLES = 2;
+
+const PRERENDER_VIEWPORT_COUNT = 3;
+
+const TAG = 'amp-next-page';
+
+/**
+ * @typedef {{
+ *   ampUrl: string,
+ *   amp: ?Object
+ * }}
+ */
+export let DocumentRef;
+
+/**
+ * Window-scoped service to handle the amp-next-page lifecycle. Handles all
+ * events and page lifecycle for the first {@code AmpNextPage} element
+ * registered. All subsequent registrations will be ignored.
+ */
+export class NextPageService {
+  constructor() {
+    /** @private {?Window} */
+    this.win_ = null;
+
+    /** @private {?Element} */
+    this.element_ = null;
+
+    /** @private {?./config.AmpNextPageConfig} */
+    this.config_ = null;
+
+    /** @private {?Element} */
+    this.separator_ = null;
+
+    /** @private {?../../../src/service/resources-impl.Resources} */
+    this.resources_ = null;
+
+    /** @private {?MultidocManager} */
+    this.multidocManager_ = null;
+
+    /** @private {number} */
+    this.nextArticle_ = 0;
+
+    /** @private {boolean} */
+    this.documentQueued_ = false;
+
+    /** @private {?../../../src/service/viewer-impl.Viewer} */
+    this.viewer_ = null;
+
+    /** @private {?../../../src/service/viewport/viewport-impl.Viewport} */
+    this.viewport_ = null;
+
+    /**
+     * @private
+     * {?../../../../src/service/position-observer/position-observer-impl.PositionObserver}
```

Comment:
FYI, this path is incorrect. It wasn't being caught by `gulp check-types` because it was split across lines.

Changing this to...
```
    /** @private {?../../../../src/service/position-observer/position-observer-impl.PositionObserver} */
```
... and running `gulp check-types` will result in this error...
```
Compiler error for check-types.js:
extensions/amp-next-page/0.1/next-page-service.js:92: WARNING - Failed to load module "../../../../src/service/position-observer/position-observer-impl.js"
     * @private {?../../../../src/service/position-observer/position-observer-impl.PositionObserver}
                  ^

extensions/amp-next-page/0.1/next-page-service.js:92: ERROR - Bad type annotation. Unknown type module$__$src$service$position_observer$position_observer_impl.PositionObserver
     * @private {?../../../../src/service/position-observer/position-observer-impl.PositionObserver}
                  ^

1 error(s), 1 warning(s)
```
... due to the incorrect path.

Fix coming up in #15247